server/src/services/backup.service.ts

```diff
@@ -0,0 +1,179 @@
+    } = this.configRepository.getEnv();
+
+    const isUrlConnection = config.connectionType === 'url';
+    const databaseParams = isUrlConnection ? [config.url] : ['-U', config.username, '-h', config.host];
+
+    try {
+      await new Promise<void>((resolve, reject) => {
+        const pgdump = this.processRepository.spawn(`pg_dumpall`, [...databaseParams, '--clean', '--if-exists'], {
+          env: { PATH: process.env.PATH, PGPASSWORD: isUrlConnection ? undefined : config.password },
+        });
+
+        const gzip = this.processRepository.spawn(`gzip`, []);
+        pgdump.stdout.pipe(gzip.stdin);
+
+        const backupFilePath = path.join(
+          StorageCore.getBaseFolder(StorageFolder.BACKUPS),
+          `immich-db-backup-${Date.now()}.sql.gz.tmp`,
+        );
+
+        const fileStream = this.storageRepository.createWriteStream(backupFilePath);
+
+        gzip.stdout.pipe(fileStream);
+
+        if (!pgdump.stderr || !pgdump.stdout) {
+          this.logger.error('Backup failed, could not spawn backup process');
+          reject('Backup failed, could not spawn backup process');
+          return;
+        }
+
+        if (!gzip.stderr || !gzip.stdout) {
+          this.logger.error('Backup failed, could not spawn gzip process');
+          reject('Backup failed, could not spawn gzip process');
+          return;
+        }
+
+        pgdump.on('error', (err) => {
+          this.logger.error('Backup failed with error', err);
+          reject(err);
+        });
+
+        gzip.on('error', (err) => {
+          this.logger.error('Gzip failed with error', err);
+          reject(err);
+        });
+
+        let pgdumpLogs = '';
+        let gzipLogs = '';
+
+        pgdump.stderr.on('data', (data) => (pgdumpLogs += data));
+        gzip.stderr.on('data', (data) => (gzipLogs += data));
+
+        pgdump.on('exit', (code) => {
+          if (code !== 0) {
+            this.logger.error(`Backup failed with code ${code}`);
+            reject(`Backup failed with code ${code}`);
+            this.logger.error(pgdumpLogs);
+            return;
+          }
+          if (pgdumpLogs) {
+            this.logger.debug(`pgdump_all logs\n${pgdumpLogs}`);
+          }
+        });
+
+        gzip.on('exit', (code) => {
+          if (code !== 0) {
+            this.logger.error(`Gzip failed with code ${code}`);
+            reject(`Gzip failed with code ${code}`);
+            this.logger.error(gzipLogs);
+            return;
+          }
+          if (pgdump.exitCode !== 0) {
+            this.logger.error(`Gzip exited with code 0 but pgdump exited with ${pgdump.exitCode}`);
+            return;
+          }
+          this.storageRepository
+            .rename(backupFilePath, backupFilePath.replace('.tmp', ''))
+            .then(() => {
+              resolve();
+            })
+            .catch((error) => {
+              this.logger.error('Backup failed with error', error);
+              reject(error);
+            });
```

Comment:
I'd prefer awaiting it and wrapping it in a try catch

Reply:
Ideally this lives outside of the promise which is only needed to interface with the callbacks, but those are done at this point in time

Reply:
Moved it just below, then we don't even need another try catch :)